File: Chapter09/C#/Node.cs
using System;
using System.Collections.Generic;
            
namespace EverydayDataStructures
{
	public class Node
	{
		public Int16 Data;
		public Node Left;
		public Node Right;
		public List<Node> Children
		{
			get
			{
				List<Node> children = new List<Node>();
				if (this.Left != null) 
				{
					children.Add (this.Left);
				}
				if (this.Right != null) 
				{
					children.Add (this.Right);
				}
				return children;
			}
		}

		public Node(Int16 data)
		{
			this.Data = data;
		}

		public bool InsertData(Int16 data)
		{
			Node node = new Node (data);
			return this.InsertNode(node);
		}

		public bool InsertNode(Node node)
		{
			if (node == null || node.Data == this.Data)
			{
				return false;
			}
			else if (node.Data < this.Data) 
			{
				if (this.Left == null) 
				{
					this.Left = node;
					return true;
				} 
				else 
				{
					return this.Left.InsertNode(node);
				}
			}
			else
			{
				if (this.Right == null) 
				{
					this.Right = node;
					return true;
				} 
				else 
				{
					return this.Right.InsertNode(node);
				}
			}
		}

		public bool Graft(Node node)
		{
			if (node == null) 
			{
				return false;
			}

			List<Node> nodes = node.ListTree();
			foreach (Node n in nodes) 
			{
				this.InsertNode(n);
			}

			return true;
		}

		public Node RemoveData(Int16 data)
		{
			Node node = new Node (data);
			return this.RemoveNode (node);
		}

		public Node RemoveNode(Node node)
		{
			if (node == null)
			{
				return null;
			}

			Node retNode;
			Node modNode;
			List<Node> treeList = new List<Node>();

			if (this.Data == node.Data)
			{
				//Root match
				retNode = new Node(this.Data);
				modNode = this;
				if (this.Children.Count == 0) 
				{
					return this; //Root has no childen
				}
			}
			else if (this.Left.Data == node.Data) 
			{
				//Match found
				retNode = new Node(this.Left.Data);
				modNode = this.Left;
			} 
			else if (this.Right.Data == node.Data)
			{
				//Match found
				retNode = new Node(this.Right.Data);
				modNode = this.Right;
			}
			else 
			{
				foreach (Node child in this.Children) 
				{
					if (child.RemoveNode(node) != null) 
					{
						return child;
					}
				}

				//No match in tree
				return null; 
			}

			//Reorder the tree
			if (modNode.Left != null) 
			{
				modNode.Data = modNode.Left.Data;
				treeList.AddRange (modNode.Left.ListTree ());
				modNode.Left = null;
			} 
			else if (modNode.Right != null) 
			{
				modNode.Data = modNode.Right.Data;
				treeList.AddRange (modNode.Right.ListTree ());
				modNode.Right = null;
			} 
			else 
			{
				modNode = null;
			}

			foreach (Node n in treeList) 
			{
				modNode.InsertNode(n);
			}

			//Finished
			return retNode;
		}

		public Node Prune(Node root)
		{
			if (root == null)
			{
				return null;
			}

			Node matchNode;
			if (this.Data == root.Data)
			{
				//Root match
				Node b = this.CopyTree();
				this.Left = null;
				this.Right = null;
				return b;
			}
			else if (this.Left.Data == root.Data) 
			{
				matchNode = this.Left;
			} 
			else if (this.Right.Data == root.Data)
			{
				matchNode = this.Right;
			}
			else 
			{
				foreach (Node child in this.Children) 
				{
					if (child.Prune(root) != null) 
					{
						return child;
					}
				}

				//No match in tree
				return null; 
			}

			Node branch = matchNode.CopyTree();
			matchNode = null;

			return branch;
		}

		public Node FindData(Int16 data)
		{
			Node node = new Node (data);
			return this.FindNode(node);
		}

		public Node FindNode(Node node)
		{
			if (node.Data == this.Data) 
			{
				return this;
			}

			foreach (Node child in this.Children) 
			{ 
				Node result = child.FindNode(node);
				if (result != null) 
				{
					return result;
				}
			}

			return null;
		}

		public Node CopyTree()
		{
			Node n = new Node(this.Data);
			if (this.Left != null)
			{
				n.Left = this.Left.CopyTree();
			}

			if(this.Right != null)
			{
				n.Right = this.Right.CopyTree();
			}
			return n;
		}
			
		public List<Node> ListTree()
		{
			List<Node> result = new List<Node>();
			result.Add(new Node(this.Data));
			foreach (Node child in this.Children)
			{
				result.AddRange(child.ListTree());
			}
			return result;
		}
	}
}

File: Chapter09/Swift/Node.swift
//
//  Node.swift
//  EverydayDataStructures-Swift
//
//  Created by Will Smith on 9/24/16.
//  Copyright © 2016 Websmiths, LLC. All rights reserved.
//

import Foundation

public class Node : Equatable
{
    public var data: Int
    public var left: Node?
    public var right: Node?

    public var children: Array<Node> {
        return [left!, right!]
    }

    public init (nodeData: Int)
    {
        data = nodeData
    }


    public func insertData(data: Int) -> Bool
    {
        return insertNode(node: Node(nodeData:data))
    }

    public func insertNode(node: Node?) -> Bool
    {
        if (node == nil)
        {
            return false
        }
        
        if ((findNode(node: node!)) != nil)
        {
            return false
        }
        else if (node!.data < data)
        {
            if (left == nil)
            {
                left = node
                return true
            }
            else
            {
                return left!.insertNode(node: node)
            }
        }
        else
        {
            if (right == node)
            {
                right = node
                return true
            }
            else
            {
                return right!.insertNode(node: node)
            }
        }
    }


    public func graft(node: Node?) -> Bool
    {
        if (node == nil)
        {
            return false
        }
        
        let nodes: Array = node!.listTree()
        for n in nodes
        {
            self.insertNode(node: n)
        }
        
        return true
    }

    public func removeData(data: Int) -> Node?
    {
        return removeNode(node: Node(nodeData:data))
    }

    public func removeNode(node: Node?) -> Node?
    {
        if (node == nil)
        {
            return nil
        }
        
        var retNode: Node
        var modNode: Node?
        var treeList = Array<Node>()
        
        if (self.data == node!.data)
        {
            //Root match
            retNode = Node(nodeData: self.data)
            modNode = self
            if (children.count == 0)
            {
                return self //Root has no childen
            }
        }
        else if (left!.data == node!.data)
        {
            //Match found
            retNode = Node(nodeData: left!.data)
            modNode = left!
        }
        else if (right!.data == node!.data)
        {
            //Match found
            retNode = Node(nodeData: right!.data)
            modNode = right!
        }
        else
        {
            for child in self.children
            {
                if (child.removeNode(node: node) != nil)
                {
                    return child
                }
            }
            
            //No match in tree
            return nil
        }
        
        //Reorder the tree
        if ((modNode!.left) != nil)
        {
            modNode!.data = modNode!.left!.data
            treeList = modNode!.left!.listTree()
            modNode!.left = nil
        }
        else if ((modNode!.right) != nil)
        {
            modNode!.data = modNode!.right!.data
            treeList = modNode!.right!.listTree()
            modNode!.right = nil
        } 
        else 
        {
            modNode = nil
        }
        
        for n in treeList
        {
            modNode!.insertNode(node: n)
        }
        
        //Finished
        return retNode
    }

    public func prune(root: Node?) -> Node?
    {
        if (root == nil)
        {
            return nil
        }
        
        var matchNode: Node?
        if (self.data == root!.data)
        {
            //Root match
            let b = self.copyTree()
            self.left = nil
            self.right = nil
            return b
        }
        else if (self.left!.data == root!.data)
        {
            matchNode = self.left!
        }
        else if (self.right!.data == root!.data)
        {
            matchNode = self.right!
        }
        else
        {
            for child in self.children
            {
                if (child.prune(root: root!) != nil)
                {
                    return child
                }
            }
            
            //No match in tree
            return nil;
        }

        let branch = matchNode!.copyTree()
        matchNode = nil

        return branch
    }
    
    public func findData(data: Int) -> Node?
    {
        return findNode(node: Node(nodeData:data))
    }
    
    public func findNode(node: Node) -> Node?
    {
        if (node == self)
        {
            return self
        }
        
        for child in children
        {
            let result = child.findNode(node: node)
            if (result != nil)
            {
                return result
            }
        }
        
        return nil
    }

    public func copyTree() -> Node
    {
        let n = Node(nodeData: self.data)
        
        if (self.left != nil)
        {
            n.left = self.left!.copyTree()
        }
        
        if(self.right != nil)
        {
            n.right = self.right!.copyTree()
        }
        
        return n
    }

    public func listTree() -> Array<Node>
    {
        var result = Array<Node>()
        result.append(self)
        for child in children
        {
            result.append(contentsOf: child.listTree())
        }
        return result
    }
}

public func == (lhs: Node, rhs: Node) -> Bool {
    return (lhs.data == rhs.data)
}
